src/main/java/org/jenkinsci/plugins/workflow/actions/TaskInfoAction.java

```diff
@@ -0,0 +1,84 @@
+package org.jenkinsci.plugins.workflow.actions;
+
+import hudson.model.InvisibleAction;
+import hudson.model.Queue;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * Records information for a {@code node} block.
+ */
+public abstract class TaskInfoAction extends InvisibleAction implements PersistentAction {
+    private static final long serialVersionUID = 1;
+
+    public enum QueueState {
+        QUEUED,
+        CANCELLED,
+        LAUNCHED,
+        UNKNOWN
+    }
+    
+    /**
+     * Gets the {@link Queue.Item} for this task, if it exists.
+     *
+     * @return The item, or null if it's not in the queue.
+     */
+    @CheckForNull
+    protected abstract Queue.Item itemInQueue();
+
+    @CheckForNull
+    public String getWhyBlocked() {
```

Comment:
I would get rid of this. Callers can call `getWhy` themselves if they want. But really they should be using `CauseOfBlockage`.